kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemPartitionIterator.java

```diff
@@ -62,15 +66,28 @@ public FileSystemIterator(int depth) throws IOException {
         dir = new Path(dir, current.get(i));
       }
 
-      try {
-        for (FileStatus stat : fs.listStatus(dir, PathFilters.notHidden())) {
-          if (stat.isDir()) {
-            // TODO: add a check here for range.couldContain(Marker)
-            dirs.add(stat.getPath().getName());
+      StorageKey key = makeKey.apply(dir);
+      boolean applies = true;
+      //if the key doesn't have any values then do not apply the predicate and assume it is ok.  The official
+      //predicate applied on the full path will exclude the details.
+      if(key.size() > 0){
+        applies = predicate.apply(key);
+      }else{
+        LOG.debug("Not applying predicate proactively because path {} does not have any key values.", dir);
+      }
+      if(applies) {
+        try {
+          for (FileStatus stat : fs.listStatus(dir, PathFilters.notHidden())) {
+            if (stat.isDir()) {
+              // TODO: add a check here for range.couldContain(Marker)
```

Comment:
That's a good question.  That TODO was there prior to the changes and I'm not sure I completely grokked the optimization.